docs: widen AGENTS.md backward-compatibility guardrail beyond PHP signatures (#66572)
woocommerce/woocommerce: AGENTS.md

Reviewer: There's some overlap with the "Hooks and filters are public contracts." paragraph we're introducing. Worth unifying?

Reviewer 2: I think in this case the overlap is okay because this paragraph doesn't just cover hooks. The later hooks paragraph is more specific.

Author: Agree with @nerrad. closing this one.

> This rule exists because WooCommerce 10.9.0 was reverted on WP Cloud: PR #64394 added a required `get_entry_count(): int` method to `FeedInterface`, fataling older WooCommerce Stripe Gateway versions that implement it. Fixed in PR #65965.

### The compatibility surface is wider than PHP signatures

WordPress exposes more contracts than class and function signatures. The following are equally binding: a change to any of them is **high-risk** and requires the same backward-compatibility impact statement in the PR description.

**Hooks and filters are public contracts.** Every `do_action` and `apply_filters` call is an interface that third-party callbacks depend on. Removing a hook, renaming it, or removing/reordering its arguments breaks every attached callback. Changing *when* or *whether* a hook fires can break consumers that depend on its timing. Additive is the safe path: append new arguments at the end, never remove or reorder existing ones. To retire a hook, fire it through `do_action_deprecated()` / `apply_filters_deprecated()` for a deprecation window instead of deleting it.

> 

Reviewer 3: Two additional sections suggested by Claude:

**Stored data is a contract**. Option names and their serialized shape, meta keys, and custom table schemas (e.g. HPOS) are read by extensions and by upgrade routines — reshaping them breaks consumers and can corrupt data on update. Add new keys/columns rather than changing existing ones; when a format must change, migrate in a versioned DB update instead of mutating in place.

**REST and Store API responses are contracts**. The fields and types a /wc/v3 or Store API endpoint returns are parsed by headless frontends, mobile apps, and integrations. Add fields, never remove or retype existing ones; version the endpoint for a breaking change.

Author: Both are real breakage classes and fit the section's thesis, but this PR is deliberately pinned to the four surfaces from the leadership steer (globals/scope, hooks/filters, multisite, install layout).

Author: Update after more thought: not committing to the fast-follow after all. The four surfaces in this PR are all gated by the section's "externally exposed" definition, but stored data and response shape have no equivalent visibility boundary — as blanket rules they would flag every option rename and every response-field tweak as high-risk. Agents treat these guardrails as hard constraints, so that reads as too strict and would dilute the section's signal. If a concrete incident class shows up here the way #64394 did for interfaces, that's the moment to add a properly scoped rule.

> **Do not assume global state.** Code can run in admin, REST, CLI, cron, webhook, and front-end contexts, and not all of them set the globals a front-end request does (`$post`, `$wp_query`, an initialized session or cart). A newly introduced read of a global, or of `WC()->…` state, in a path reachable outside a standard request is a fatal or a silent misbehavior in the contexts that do not set it. Guard availability explicitly (`function_exists`, `class_exists`, `isset`, `did_action`) rather than assuming it.

Author: Applied in 73f922a8ac - the four guard types are now named for what they each cover, with WC() + component initialization called out explicitly before dereferencing.

> **Do not assume single-site.** Multisite changes where data lives: site-scoped vs network-scoped options (`get_option` vs `get_site_option`), per-site tables, user roles and capabilities, and upload paths all differ. A change that reads or writes site state must state in its PR whether it behaves correctly under multisite — and if it was not tested there, say so explicitly.

Author: "and if it was not tested there, say so explicitly." - too strong?

Reviewer 2: I'd say leave it for now.

> **Do not assume install layout.** WordPress runs in subdirectories, with relocated `wp-content`, and behind reverse proxies. Never build paths or URLs by concatenation from the domain root; derive them (`plugins_url()`, `plugin_dir_path()`, `wp_upload_dir()`, and mind the `home_url()` vs `site_url()` distinction). A path that works on a root install and breaks elsewhere is a compatibility bug, not an edge case.

Reviewer 2: ```suggestion
**Do not assume install layout.** WordPress could be configured to run in a subdirectory, with relocated `wp-content`, and behind reverse proxies. Never build paths or URLs by concatenation from the domain root; derive them (`plugins_url()`, `plugin_dir_path()`, `wp_upload_dir()`, and mind the `home_url()` vs `site_url()` distinction). A path that works on a root install and breaks elsewhere is a compatibility bug, not an edge case.
```

Author: Applied in cae690c02d.

> ### Before changing any exposed surface (agent checklist)

Reviewer: Is "exposed" clear enough? Is the intent to be "public", but to also cover "implicitly public" surface?

Author: changed it to be more explicit "public or externally exposed": ccd4610047.

1. Identify the contract you are touching: signature, hook, global/scope expectation, site topology, or install layout.
2. Assume unseen consumers. You cannot enumerate third-party code; if the surface is reachable from outside this plugin, someone consumes it.
3. Prefer the additive path (new optional method, appended hook argument, new symbol + deprecation) over changing what exists.

> 4. State the impact in the PR description: what changed, who could consume it, and why it is safe or what the deprecation path is.

Reviewer: I really like these:

* Assume unseen consumers
* If reachable, someone consumes it
* Be explicit about who could consume it

Should we implement them in the above guidelines?

Author: They're in the prose in scoped form, the section opening already says "When in doubt, assume it is exposed and state the BC impact." The checklist states the unconditional form ("assume unseen consumers") because it's already gated to changes on an exposed surface. 

If you think the lead should carry "assume unseen consumers" verbatim I can change it, but I'd keep it this way.

Author: closing for now

> 5. If you cannot establish the impact, stop and flag it in the PR instead of proceeding.

Reviewer 2: What does "instead of proceeding" mean here? By the time the PR is created the code has already been written? Perhaps something like this instead:

```suggestion
5. If you cannot establish the impact, flag it to the user as needing review.
```

Author: Applied in 95a02b3bbf, with one adjustment, I kept `stop`. The incoherent part was pointing the escalation at a PR that doesn't exist yet; the authoring-time halt (don't make the change) is worth keeping. 

Now it reads: "If you cannot establish the impact, stop and flag it to the user as needing review."
